channeldb: add optional migration to prune revocation logs

yyforyongyu opened the pull request:
This is a follow-up PR on #6347.

This PR introduces an optional migration that's fault-tolerant. Since the potential scale is quite large, to prevent the OOM issue we need to perform the update in different db transactions. We cannot safely assume the process won't be interrupted, thus we need to be able to pick up where we left off and continue the migration if it does happen.

This PR also introduces `optional meta` that's similar to `meta` but saves the optional meta info. We need to separate the management of the two because when we don't, after a future mandatory migration is applied, the user would have trouble applying the optional migration.

yyforyongyu commented:
Here're some benchmark results, maybe there's something wrong with the setup but the migration is really slow.

```
Benchmark for 2000 logs

    2000     214401347 ns/op     7181999 B/op      92167 allocs/op
PASS
ok      github.com/lightningnetwork/lnd/channeldb/migration26   576.254s



Benchmark for 5000 logs

    5000     250081490 ns/op     9413385 B/op     125960 allocs/op
PASS
ok      github.com/lightningnetwork/lnd/channeldb/migration26   1549.976s
```

yyforyongyu commented on channeldb/migration26/migration.go:
```diff
@@ -0,0 +1,275 @@
+// transaction. In the worst case, each old revocation log is roughly 0.75
+// MB(747573 bytes). 1000 records would consume 0.75 GB of ram, which is
+// feasible for a modern machine.
+const recordsPerTx = 1000
```
This figure might be way toooo conservative. I'm thinking instead of tracking the number of records, we can use `runtime.ReadMemStats` to monitor and cap the memory usage.

yyforyongyu replied:
Decided to convert the old large commitments into revocation logs first to save more memory. However, as the benchmark has shown, the CPU was not bounded by this memory usage at all. Rather, most of the CPU time was taken by `DeriveCommitmentKeys` that's used to calculate the output indexes.

yyforyongyu commented on channeldb/migration26/migration.go:
```diff
@@ -0,0 +1,275 @@
+		}
+
+		err = kvdb.Update(db, func(tx kvdb.RwTx) error {
+			finished, err = processMigration(tx)
```
Maybe log something here to show the progress?

yyforyongyu replied:
Added a progress.

yyforyongyu commented:
Decided to put the script computation inside a goroutine, plus a bug fixed in `iterator`, the speed is much faster now.
```
Building test data for 1000 logs...
Finished building test data, migrating...
    1000        195708 ns/op       11526 B/op        178 allocs/op
PASS
ok      github.com/lightningnetwork/lnd/channeldb/migration26   0.626s

Building test data for 5000 logs...
Finished building test data, migrating...
    5000         82152 ns/op       11615 B/op        177 allocs/op
PASS
ok      github.com/lightningnetwork/lnd/channeldb/migration26   0.873s


Building test data for 1000000 logs...
Finished building test data, migrating...
 1000000         56819 ns/op       11370 B/op        175 allocs/op
PASS
ok      github.com/lightningnetwork/lnd/channeldb/migration26   71.224s
```

Roasbeef commented on channeldb/migration26/iterator.go:
```diff
@@ -0,0 +1,257 @@
+	return chanBucket, nil
+}
+
+// findNextMigrateHeight finds the next commit height that's not migrated. It
```
Very cool, so this is intended to enable us to not have to commit the entire migration in a single transaction and to support incremental restart?

Roasbeef commented on channeldb/migration26/iterator.go:
```diff
@@ -0,0 +1,257 @@
+
+	// Iterate the buckets. If we received an exit signal, return the
+	// locator.
+	err := iterateBuckets(openChanBucket, nil, cb)
```
Shouldn't we be passing in a specific node's bucket here rather than the entire open channel bucket? As we need to know the last location for each channel, not the last _global_ key we migrated?

yyforyongyu replied:
Actually it's designed to locate the global key. Maybe I should've named the functions better, but `findNextMigrateHeight` is designed to find the next key within the scope of a *channel*, and `locateNextUpdateNum` is to find the *global* position.

Suppose we restart a previously interrupted migration, we need to,
```
-> iterate the node's buckets until,
    -> iterate the chainhash buckets until,
        -> iterate the channel buckets until,
            -> a channel's next revocation log height is located, which signals the exit.
```

In other words a restart means we need to find out which channel inside which node has unfinished jobs and continue the migration there.

Roasbeef commented on channeldb/migration26/iterator.go:
```diff
@@ -0,0 +1,257 @@
+			})
+	}
+
+	return iterator(openChanBucket, l.nodePub, func(k1, v []byte) error {
```
Cool pattern here!

Roasbeef commented on channeldb/migration26/migration.go:
```diff
@@ -0,0 +1,499 @@
+
+			logEntrykey := mig24.MakeLogKey(entry.commitHeight)
+			err = logBucket.Put(logEntrykey[:], b.Bytes())
+
```
Style nit: extra new line.

yyforyongyu replied:
Fixed!

Roasbeef reviewed (commented):
Completed a _very_ cursory pass over to get a better grasp of the proposed approach and also how the new optional migration scaffolding has been added. Great job on the set of test cases! They're very comprehensive, and I also like that we started to benchmark the migrations as well to get a handle on how long it'll take a canned node to apply the migration. Ideally in the near future we never need to do such manual migrations again, but this feature is certainly urgently needed. 

Will circle back with a deeper review, and also some test runs (the `--dry-run-migration` flag can be super useful here). Speaking of `--dry-run-migration`, as is it the migration in this PR isn't strictly compatible with that, as the PR will do smaller transactions to commit the progress within, vs the existing approach where one mega transaction is used for everything.

yyforyongyu commented:
> Speaking of --dry-run-migration, as is it the migration in this PR isn't strictly compatible with that, as the PR will do smaller transactions to commit the progress within, vs the existing approach where one mega transaction is used for everything.

Yeah that's indeed an issue as the migration relies on what's already written on disk to decide its next move. Maybe we could delete the new buckets if we are in dry-run mode? While it complicates things more, I think we can leverage the reset callback inside `kvdb.Update` to wipe out data that's previously written.

Crypt-iQ commented on channeldb/migration26/migration.go:
```diff
@@ -0,0 +1,498 @@
+
+	for locator, logs := range entryMap {
+		// Find the channel bucket.
+		chanBucket, err := locator.locateChanBucket(openChanBucket)
```
Pretty sure the locator's height needs to be copied, the cursor returns read-only values directly from bbolt's mmap, which can get resized here since mutations (insertions) are happening here. Using old mmap values either gives corrupt values or a panic

yyforyongyu replied:
We don't use the locator's height here, maybe I should've separated the two types of locators.

yyforyongyu replied:
What's used here are `nodePub`, `chainhash` and `fundingOutpoint`. Since we are not touching any of the upper-level buckets, and only working on the lowest-nested bucket that's located by these three keys, I'm assuming it's fine as the insertions happen in a new block of memory/disk space?

Also I think the `updateLocator`'s pointer is not managed by bbolt and, when we create it, we just copy the values read from that cursor.

Meanwhile I agree that it's better to use `updateLocator` instead of `*updateLocator` everywhere to avoid concerns like such.

lightninglabs-deploy commented:
@yyforyongyu, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu, remember to re-request review from reviewers when ready

TonyGiorgio commented:
Ran this PR on my 30GB channel.db that wouldn't compact down any further and it was able to compact it down to 129MB, giving a thumbs up, thanks :+1:

lightninglabs-deploy commented:
@yyforyongyu, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu, remember to re-request review from reviewers when ready

Roasbeef commented on channeldb/migration29/iterator.go:
```diff
@@ -0,0 +1,257 @@
+	cb := func(chanBucket kvdb.RwBucket, l *updateLocator) error {
+		locator = l
+		updateNum := findNextMigrateHeight(chanBucket)
+
```
I think we should log the height here, so we can use it to print some more progress during the migration. Otherwise, users may just see it sitting there for minutes and wonder what's going on.

yyforyongyu replied:
Added more verbose logs and debug logs.

Roasbeef commented on channeldb/migration29/migration.go:
```diff
@@ -0,0 +1,498 @@
+		return err
+	}
+
+	log.Infof("Total un-migrated logs: %d", total)
```
Ah ok , so this _does_ have some logging. However we never registered this logger so nothing is printed out atm.

Roasbeef replied:
```diff
diff --git a/channeldb/log.go b/channeldb/log.go
index 42b55cc..658b53d 100644
--- a/channeldb/log.go
+++ b/channeldb/log.go
@@ -8,6 +8,7 @@ import (
        "github.com/lightningnetwork/lnd/channeldb/migration13"
        "github.com/lightningnetwork/lnd/channeldb/migration16"
        "github.com/lightningnetwork/lnd/channeldb/migration24"
+       "github.com/lightningnetwork/lnd/channeldb/migration29"
        "github.com/lightningnetwork/lnd/channeldb/migration_01_to_11"
        "github.com/lightningnetwork/lnd/kvdb"
 )
@@ -38,5 +39,6 @@ func UseLogger(logger btclog.Logger) {
        migration13.UseLogger(logger)
        migration16.UseLogger(logger)
        migration24.UseLogger(logger)
+       migration29.UseLogger(logger)
        kvdb.UseLogger(logger)
 }
```

yyforyongyu replied:
cool somehow missed, added.

Roasbeef commented on channeldb/migration29/migration.go:
```diff
@@ -0,0 +1,498 @@
+			log.Infof("Migrating revocation logs finished!")
+			break
+		}
+
```
As is, the migration isn't actually _easily_ interrupt-able (which seems to be a goal). To achieve this, we need to pass in a quit channel that'll be hooked up to the normal signal.

Roasbeef commented on channeldb/migration29/iterator.go:
```diff
@@ -0,0 +1,257 @@
+// locateNextUpdateNum returns a locator that's used to start our migration. A
+// nil locator means the migration has been finished.
+func locateNextUpdateNum(openChanBucket kvdb.RwBucket) (*updateLocator, error) {
+	locator := &updateLocator{}
```
Shouldn't this be scoped to a given channel or peer pub key?

Roasbeef commented on channeldb/migration29/migration.go:
```diff
@@ -0,0 +1,498 @@
+// revocation buckets.
+func deleteOldBuckets(db kvdb.Backend) error {
+	// locators records all the chan buckets found in the database.
+	locators := make([]*updateLocator, 0)
```
Can be slightly simplified to: `var locators []*updateLoactor`.

yyforyongyu replied:
Done.

Roasbeef commented on channeldb/migration29/migration.go:
```diff
@@ -0,0 +1,498 @@
+		defer wg.Done()
+
+		// Find the output indexes.
+		ourIndex, theirIndex, err := findOutputIndexes(chanState, &c)
```
Was running a migration, the first 67% or so was pretty quick, but then I noticed it really start to get bogged down. I took a profile, and realized that this is the function that soaks up the most CPU: 
![Screen Shot 2022-06-30 at 7 39 34 PM](https://user-images.githubusercontent.com/998190/176812943-ee8e8c41-f708-499b-a41b-63d15bcd29da.png)

Roasbeef replied:
One idea here would be to make a goroutine worker pool and submit this job to them.

Roasbeef commented on channeldb/db.go:
```diff
@@ -1367,6 +1405,61 @@ func (d *DB) syncVersions(versions []version) error {
+// migrations will be applied.
+//
+// NOTE: only support the prune_revocation_log optional migration atm.
+func (d *DB) applyOptionalVersions(cfg OptionalMiragtionConfig) error {
```
So we'd thread through that quit channel _here_ to make the migration resumable.

yyforyongyu replied:
So I guess to achieve this, we'll pass the `ctx` used in `BuildDatabase` through `BuildDatabase -> CreateWithBackend -> applyOptionalVersions -> version.migration` and catch the cancel signal there?

Roasbeef commented on channeldb/migration29/lnwallet.go:
```diff
@@ -0,0 +1,361 @@
+
+	// With the commitment point generated, we can now derive the king ring
+	// which will be used to generate the output scripts.
+	keyRing := DeriveCommitmentKeys(
```
If we can perform this same check w/o needing to derive the commitment keys, then we should be able to speed up the migration significantly.

yyforyongyu replied:
Not sure I understand. I think we need to derive the king ring here to compute the scripts so we can derive the output indexes?

Roasbeef commented:
So I tried to profile an active run to see what all the goroutines were up to, but I think the usage of internal closures is messing things up, rather than see all the stack traces, all I see instead is:
```
1090 @ 0x4710a1

```

Roasbeef commented:
Will do a deeper dive tomorrow, but something is up w/ the progress logging, and due to the issue I mentioned above, I'm not able to get a proper stack trace or profile to see what's up:
```
2022-06-30 23:10:45.215 [INF] CHDB: Performing database optional migration: prune revocation log
2022-06-30 23:10:45.216 [INF] CHDB: Migrating revocation logs, might take a while...
2022-06-30 23:10:45.224 [INF] CHDB: Total un-migrated logs: 59661
2022-06-30 23:11:21.543 [INF] CHDB: Migration progress: 33.523%, still have: 39661
2022-06-30 23:11:58.590 [INF] CHDB: Migration progress: 67.045%, still have: 19661
```

It's been sitting there for 30 mins or so now as well. 

It also picked back up from zero when it was restarted prior: 
```
2022-06-30 22:16:44.410 [INF] CHDB: Performing database optional migration: prune revocation log
2022-06-30 22:16:44.411 [INF] CHDB: Migrating revocation logs, might take a while...
2022-06-30 22:16:44.419 [INF] CHDB: Total un-migrated logs: 59661
2022-06-30 22:17:19.624 [INF] CHDB: Migration progress: 33.523%, still have: 39661
2022-06-30 22:17:54.893 [INF] CHDB: Migration progress: 67.045%, still have: 19661
2022-06-30 23:10:17.330 [INF] LTND: Received interrupt
2022-06-30 23:10:17.332 [INF] LTND: Shutting down...
2022-06-30 23:10:17.333 [INF] LTND: Gracefully shutting down.
```

FWIW, this is a _mega_ old node. I think it turned 5 earlier this year...

lightninglabs-deploy commented:
@yyforyongyu, remember to re-request review from reviewers when ready

yyforyongyu commented:
Updates from last push,
- added more verbose logging
- fixed the case where both new and old logs exist
- added a defense check prior to the deletion of old buckets

yyforyongyu commented:
> It's been sitting there for 30 mins or so now as well. FWIW, this is a pretty tiny box (1 vCPU), 2 GB of RAM.

Here's what went wrong,
- previously the migration had a defense check when it tried to locate the next migration height. When the height key found in the new bucket was greater, the bucket was considered corrupted and the migration would be applied again.
- this would be fine prior to `v0.15.0` as no new logs were expected, hence the height saved in the new buckets won't be greater than the old un-migrated heights.
- with `v0.15.0` things are different, new logs are saved to the new buckets already, and their heights must be greater.
- The above case will cause an infinite loop where 1) new heights are greater -> 2) redo the current migration.
- This is now fixed by taking the above situation into account.

Roasbeef commented:
@yyforyongyu great work! Confirmed that I was able to successfully rerun the migration locally. This time around the entire thing was done in less than a minute or so. Now that this is fixed, I'll circle back again re review+testing for this PR.

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu, remember to re-request review from reviewers when ready

Roasbeef reviewed (approved):
LGTM 🧜‍♀️

The latest diff fixed the issue on my testnet node, and I've run this on one of my mainnet nodes successfully!

bhandras commented on channeldb/meta.go:
```diff
@@ -80,3 +88,92 @@ func putDbVersion(metaBucket kvdb.RwBucket, meta *Meta) error {
+		}
+
+		return nil
+	}, func() {})
```
Even though we don't know of people running nodes on etcd if they do, it's safer if we reset any external state properly, like in this current block we update the `om.Versions` map which is external to the closure.

bhandras reviewed (approved):
LGTM :tada:  Some parts of the code are not known by me, so tried to do a more code oriented review, however there's not much to add, beautiful work :1st_place_medal:

C-Otto commented:
For those who are curious: I was able to run the migration (via the `TestFoo` script above) for a copy of my channel.db (16 GByte, c-otto.de). It took 249 seconds on my Intel i5 12600k for around 3.7 million logs.

rkfg commented:
Out of curiosity, what was the size of the database after migration? My db is only 3.8 Gb so probably compacting it isn't an emergency, I'll wait for more reports.

C-Otto commented:
```
2022-08-05 17:29:52.841 [INF] CHDB: DB compaction of /home/cotto/.lnd/data/graph/mainnet/channel.db successful, 16960901120 -> 4039794688 bytes (gain=4.20x)
```

rkfg commented:
Nice, though not as drastic as in https://github.com/lightningnetwork/lnd/pull/6469#issuecomment-1146455568, I wonder why... has something changed since that? Maybe the final version of the PR prunes less data?

C-Otto commented:
My database didn't include a lot of other cruft (that a regular compaction would remove). Also, I've been using lnd 0.15 for a while now, so several newer channels don't benefit from this patch.

yyforyongyu commented:
> I wonder why... has something changed since that? Maybe the final version of the PR prunes less data?

Nope it's the same. You may not see such a gain because you probably have lots of payments and invoices, which can take up some disk space.
